server/src/test/java/org/elasticsearch/common/util/ObjectObjectPagedHashMapTests.java

```diff
@@ -0,0 +1,74 @@
+/*
+ * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+ * or more contributor license agreements. Licensed under the Elastic License
+ * 2.0 and the Server Side Public License, v 1; you may not use this file except
+ * in compliance with, at your election, the Elastic License 2.0 or the Server
+ * Side Public License, v 1.
+ */
+
+package org.elasticsearch.common.util;
+
+import org.apache.lucene.util.BytesRef;
+import org.elasticsearch.common.settings.Settings;
+import org.elasticsearch.common.unit.ByteSizeValue;
+import org.elasticsearch.indices.breaker.NoneCircuitBreakerService;
+import org.elasticsearch.test.ESTestCase;
+
+import java.util.HashMap;
+import java.util.Map;
+
+public class ObjectObjectPagedHashMapTests extends ESTestCase {
+
+    private BigArrays mockBigArrays() {
+        return new MockBigArrays(new MockPageCacheRecycler(Settings.EMPTY), new NoneCircuitBreakerService());
+    }
+
+    public void testDuel() {
+        final Map<BytesRef, Object> map1 = new HashMap<>();
+        final ObjectObjectPagedHashMap<BytesRef, Object> map2 = new ObjectObjectPagedHashMap<>(
+            randomInt(42),
+            0.6f + randomFloat() * 0.39f,
+            mockBigArrays()
+        );
+        final int maxKey = randomIntBetween(1, 10000);
+        BytesRef[] bytesRefs = new BytesRef[maxKey];
+        for (int i = 0; i < maxKey; i++) {
+            bytesRefs[i] = randomBytesRef();
+        }
+        final int iters = scaledRandomIntBetween(10000, 100000);
+        for (int i = 0; i < iters; ++i) {
+            final boolean put = randomBoolean();
+            final int iters2 = randomIntBetween(1, 100);
+            for (int j = 0; j < iters2; ++j) {
+                final BytesRef key = bytesRefs[random().nextInt(maxKey)];
+                if (put) {
+                    final Object value = new Object();
+                    assertSame(map1.put(key, value), map2.put(key, value));
+                } else {
+                    assertSame(map1.remove(key), map2.remove(key));
+                }
+                assertEquals(map1.size(), map2.size());
+            }
+        }
+        for (int i = 0; i < maxKey; i++) {
+            assertSame(map1.get(bytesRefs[i]), map2.get(bytesRefs[i]));
+        }
+        final Map<BytesRef, Object> copy = new HashMap<>();
+        for (ObjectObjectPagedHashMap.Cursor<BytesRef, Object> cursor : map2) {
+            copy.put(cursor.key, cursor.value);
+        }
+        map2.close();
+        assertEquals(map1, copy);
+    }
```

Comment:
What about doing it with the cranky breaker to make sure it'll handle failures?

Reply:
Done, unfortunately I needed to change the visibility on one of the constructor in `MockBigArrays`, wdyt?

Reply:
It's all good. I could swear we do this sort of thing in ESQL already. But I don't mind the visibility change. It'd be nice if you could stick javadocs on the ctors now that there are a bunch visible.